File: crates/pamoja-profile/src/node.rs
//! The ready-to-run node a profile assembles around real components.

use core::time::Duration;

use pamoja_codec::Codec;
use pamoja_core::{Actuator, Result, Sensor, Transport};
use pamoja_power::PowerMode;

use crate::{Controller, Policy, Profile, Reaction};

/// An actuator that accepts and ignores commands.
///
/// Profiles that only observe - such as a well-level monitor - have no output to
/// drive. [`Node::monitor`] wires this in their place so a node has one uniform
/// shape whether or not it switches an actuator.
#[derive(Clone, Copy, Debug, Default)]
pub struct NoActuator;

impl Actuator for NoActuator {
    type Command = bool;

    async fn apply(&mut self, _command: bool) -> Result<()> {
        Ok(())
    }
}

/// A profile assembled around the components that make it run.
///
/// The node is the thin shell that ties a [`Profile`]'s decision logic to real I/O.
/// Each [`tick`](Node::tick) reads the sensor, runs the node's policy, drives the
/// actuator when the policy calls for it, and publishes the reading over the
/// transport with the supplied codec. The control math lives in `pamoja-kit`, the
/// power schedule in `pamoja-power`, and the wire format in the codec, so the node
/// adds composition, not behavior.
///
/// [`Node::new`] runs the profile's [`Controller`], which reads real-world `f32`
/// units (degrees, percent, liters), the form the `pamoja-kit` controllers expect; a
/// driver is responsible for calibrating raw counts into those units before the node
/// sees them, or [`Sensor::map`] selects one. [`Node::with_policy`] runs a
/// [`Policy`] of your own instead, over whatever the sensor reads and whatever the
/// actuator takes.
///
/// # Examples
///
/// ```
/// use pamoja_codec::CborCodec;
/// use pamoja_core::{Actuator, Result, Sensor, Transport};
/// use pamoja_loopback::{LoopbackBroker, LoopbackTransport};
/// use pamoja_profile::{Node, Profile};
///
/// struct Probe(f32);
/// impl Sensor for Probe {
///     type Reading = f32;
///     async fn read(&mut self) -> Result<f32> {
///         Ok(self.0)
///     }
/// }
///
/// struct Cooler;
/// impl Actuator for Cooler {
///     type Command = bool;
///     async fn apply(&mut self, _on: bool) -> Result<()> {
///         Ok(())
///     }
/// }
///
/// # async fn run() -> Result<()> {
/// let broker = LoopbackBroker::new();
/// let mut link = LoopbackTransport::new(broker);
/// link.connect().await?;
///
/// // A warm fridge, assembled straight from its profile.
/// let mut node = Node::new(Profile::vaccine_fridge_monitor(), Probe(9.0), Cooler, link, CborCodec);
/// let reaction = node.tick().await?;
/// assert_eq!(reaction.actuator, Some(true)); // the cooler runs
/// assert!(reaction.alert.is_some()); // and 9 C is a spoilage excursion
/// # Ok(())
/// # }
/// ```
pub struct Node<S, A, T, C, P = Controller> {
    profile: Profile,
    policy: P,
    sensor: S,
    actuator: A,
    transport: T,
    codec: C,
}

impl<S, A, T, C> Node<S, A, T, C> {
    /// Assembles a node from a profile and the components that drive it.
    ///
    /// # Arguments
    ///
    /// * `profile` - the profile to assemble; its policy becomes the node's controller.
    /// * `sensor` - the source of readings.
    /// * `actuator` - the output the controller switches.
    /// * `transport` - the link readings are published over; expected to be connected.
    /// * `codec` - the wire format readings are encoded with.
    ///
    /// # Returns
    ///
    /// A node ready to [`tick`](Node::tick).
    pub fn new(profile: Profile, sensor: S, actuator: A, transport: T, codec: C) -> Self {
        let controller = profile.controller();
        Node::with_policy(profile, controller, sensor, actuator, transport, codec)
    }
}

impl<S, A, T, C, P> Node<S, A, T, C, P> {
    /// Assembles a node around a policy of your own.
    ///
    /// The profile still names the topic and the power schedule; the policy decides
    /// each reading, so a node can read a driver's whole measurement and command an
    /// actuator that takes more than on and off. A policy a [`PolicyRegistry`] resolved
    /// from the profile's own control kind goes here too.
    ///
    /// # Arguments
    ///
    /// * `profile` - the profile to assemble.
    /// * `policy` - what decides each reading.
    /// * `sensor` - the source of readings.
    /// * `actuator` - the output the policy commands.
    /// * `transport` - the link readings are published over; expected to be connected.
    /// * `codec` - the wire format readings are encoded with.
    ///
    /// # Returns
    ///
    /// A node ready to [`tick`](Node::tick).
    ///
    /// # Examples
    ///
    /// ```
    /// use pamoja_codec::CborCodec;
    /// use pamoja_core::{Actuator, Result, Sensor, Transport};
    /// use pamoja_loopback::{LoopbackBroker, LoopbackTransport};
    /// use pamoja_profile::{Alert, Node, Policy, Profile, Reaction};
    /// use serde::{Deserialize, Serialize};
    ///
    /// // A driver reads two quantities at once; the policy decides on both.
    /// #[derive(Clone, Copy, Serialize, Deserialize)]
    /// struct Soil {
    ///     moisture: f32,
    ///     temperature: f32,
    /// }
    ///
    /// struct Probe;
    /// impl Sensor for Probe {
    ///     type Reading = Soil;
    ///     async fn read(&mut self) -> Result<Soil> {
    ///         Ok(Soil { moisture: 22.0, temperature: 1.5 })
    ///     }
    /// }
    ///
    /// struct Valve;
    /// impl Actuator for Valve {
    ///     type Command = bool;
    ///     async fn apply(&mut self, _open: bool) -> Result<()> {
    ///         Ok(())
    ///     }
    /// }
    ///
    /// // Water only when the bed is dry and not about to freeze.
    /// struct DripPolicy;
    /// impl Policy for DripPolicy {
    ///     type Reading = Soil;
    ///     type Command = bool;
    ///     fn evaluate(&mut self, soil: &Soil) -> Reaction {
    ///         let frost = soil.temperature < 2.0;
    ///         Reaction {
    ///             actuator: Some(soil.moisture < 25.0 && !frost),
    ///             alert: frost.then_some(Alert::Custom { code: "FrostRisk", value: soil.temperature }),
    ///         }
    ///     }
    /// }
    ///
    /// # async fn run() -> Result<()> {
    /// let mut link = LoopbackTransport::new(LoopbackBroker::new());
    /// link.connect().await?;
    /// let profile = Profile::irrigation_node();
    /// let mut node = Node::with_policy(profile, DripPolicy, Probe, Valve, link, CborCodec);
    /// let reaction = node.tick().await?;
    /// assert_eq!(reaction.actuator, Some(false)); // dry, but frost is closer
    /// assert_eq!(reaction.alert.map(Alert::kind), Some("FrostRisk"));
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// [`PolicyRegistry`]: crate::PolicyRegistry
    pub fn with_policy(
        profile: Profile,
        policy: P,
        sensor: S,
        actuator: A,
        transport: T,
        codec: C,
    ) -> Self {
        Self {
            profile,
            policy,
            sensor,
            actuator,
            transport,
            codec,
        }
    }

    /// Returns the profile this node was assembled from.
    ///
    /// # Returns
    ///
    /// A reference to the node's [`Profile`].
    pub fn profile(&self) -> &Profile {
        &self.profile
    }

    /// Returns the policy deciding this node's readings.
    ///
    /// # Returns
    ///
    /// A reference to the node's policy.
    pub fn policy(&self) -> &P {
        &self.policy
    }

    /// Returns the power mode and wait interval for the next cycle.
    ///
    /// This assembles the profile's [`PowerSchedule`](crate::PowerSchedule) into a
    /// `pamoja-power` governor: as the battery drains the interval stretches, and a
    /// charging panel eases the node back toward its active cadence. The node never
    /// sleeps; the caller waits the
    /// returned [`Duration`] before the next [`tick`](Node::tick), so timing stays
    /// outside the node and the decision logic remains synchronous and testable.
    ///
    /// # Arguments
    ///
    /// * `soc` - the battery state of charge in `[0.0, 1.0]`.
    /// * `charging` - whether the panel is currently delivering charge.
    ///
    /// # Returns
    ///
    /// The [`PowerMode`] to run in and how long to wait before the next cycle.
    pub fn schedule(&self, soc: f32, charging: bool) -> (PowerMode, Duration) {
        let plan = self.profile.power.plan();
        let mode = plan.mode_while_charging(soc, charging);
        (mode, plan.interval_for(mode))
    }
}

impl<S, T, C> Node<S, NoActuator, T, C> {
    /// Assembles a node for a profile that observes without driving an output.
    ///
    /// Wires a [`NoActuator`] in place of a real output, so a monitoring profile such
    /// as [`well_level`](Profile::well_level) reads and publishes with the same shape
    /// as a controlling one.
    ///
    /// # Arguments
    ///
    /// * `profile` - the profile to assemble.
    /// * `sensor` - the source of readings.
    /// * `transport` - the link readings are published over; expected to be connected.
    /// * `codec` - the wire format readings are encoded with.
    ///
    /// # Returns
    ///
    /// A node ready to [`tick`](Node::tick), with no output to switch.
    pub fn monitor(profile: Profile, sensor: S, transport: T, codec: C) -> Self {
        Node::new(profile, sensor, NoActuator, transport, codec)
    }
}

impl<S, A, T, C, P> Node<S, A, T, C, P>
where
    S: Sensor,
    P: Policy<Reading = S::Reading>,
    P::Command: Clone,
    A: Actuator<Command = P::Command>,
    T: Transport,
    C: Codec<S::Reading>,
{
    /// Runs one read-decide-act-publish cycle.
    ///
    /// Reads the sensor, evaluates the node's policy, applies the resulting command
    /// to the actuator when the policy calls for one, and publishes the reading to
    /// the profile's topic.
    ///
    /// # Returns
    ///
    /// The [`Reaction`] the policy produced: the command that was applied (if any)
    /// and any alert the reading raised.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Io`](pamoja_core::Error::Io) or
    /// [`Error::Closed`](pamoja_core::Error::Closed) if the sensor read or the
    /// actuator command fails, [`Error::Codec`](pamoja_core::Error::Codec) if the
    /// reading cannot be encoded, and [`Error::Transport`](pamoja_core::Error::Transport)
    /// or [`Error::Closed`](pamoja_core::Error::Closed) if the publish fails.
    pub async fn tick(&mut self) -> Result<Reaction<P::Command>> {
        let reading = self.sensor.read().await?;
        let reaction = self.policy.evaluate(&reading);
        if let Some(command) = reaction.actuator.clone() {
            self.actuator.apply(command).await?;
        }
        let payload = self.codec.encode(&reading)?;
        self.transport.send(&self.profile.topic, &payload).await?;
        Ok(reaction)
    }
}

#[cfg(test)]
mod tests {
    use std::sync::{Arc, Mutex};

    use pamoja_codec::CborCodec;
    use pamoja_core::Error;
    use pamoja_core::Receive;
    use pamoja_loopback::{LoopbackBroker, LoopbackTransport};
    use pamoja_power::PowerMode;

    use super::*;
    use crate::Alert;

    // A sensor that plays back a fixed list of readings, then reports closed.
    struct ScriptedSensor {
        readings: std::vec::IntoIter<f32>,
    }

    impl ScriptedSensor {
        fn new(readings: Vec<f32>) -> Self {
            Self {
                readings: readings.into_iter(),
            }
        }
    }

    impl Sensor for ScriptedSensor {
        type Reading = f32;

        async fn read(&mut self) -> Result<f32> {
            self.readings.next().ok_or(Error::Closed)
        }
    }

    // An actuator that records every command it is given.
    #[derive(Clone)]
    struct Recording<C = bool>(Arc<Mutex<Vec<C>>>);

    impl<C: Send> Actuator for Recording<C> {
        type Command = C;

        async fn apply(&mut self, command: C) -> Result<()> {
            self.0.lock().expect("commands lock").push(command);
            Ok(())
        }
    }

    async fn connected_pair(filter: &str) -> (LoopbackTransport, LoopbackTransport) {
        let broker = LoopbackBroker::new();
        let mut gateway = LoopbackTransport::new(broker.clone());
        let mut link = LoopbackTransport::new(broker);
        gateway.connect().await.expect("gateway connect");
        link.connect().await.expect("link connect");
        gateway.subscribe(filter).await.expect("subscribe");
        (gateway, link)
    }

    #[tokio::test]
    async fn tick_reads_actuates_and_publishes() {
        let (mut gateway, link) = connected_pair("cold-chain/#").await;
        let commands = Arc::new(Mutex::new(Vec::new()));
        let mut node = Node::new(
            Profile::vaccine_fridge_monitor(),
            ScriptedSensor::new(vec![9.0]),
            Recording(commands.clone()),
            link,
            CborCodec,
        );

        let reaction = node.tick().await.expect("tick");
        assert_eq!(reaction.actuator, Some(true));
        assert!(reaction.alert.is_some());
        assert_eq!(*commands.lock().expect("commands"), vec![true]);

        let message = gateway.recv().await.expect("recv").expect("a reading");
        assert_eq!(message.topic, "cold-chain/fridge/temperature");
        let reading: f32 = CborCodec.decode(&message.payload).expect("decode");
        assert_eq!(reading, 9.0);
    }

    #[tokio::test]
    async fn monitor_publishes_without_an_actuator() {
        let (mut gateway, link) = connected_pair("water/#").await;
        let mut node = Node::monitor(
            Profile::well_level(),
            ScriptedSensor::new(vec![3.2]),
            link,
            CborCodec,
        );

        let reaction = node.tick().await.expect("tick");
        assert_eq!(reaction.actuator, None);

        let message = gateway.recv().await.expect("recv").expect("a reading");
        assert_eq!(message.topic, "water/well/level");
        let reading: f32 = CborCodec.decode(&message.payload).expect("decode");
        assert_eq!(reading, 3.2);
    }

    // A sensor reading two quantities at once, as a driver for a combined part does.
    struct Climate(f32, f32);

    impl Sensor for Climate {
        type Reading = (f32, f32);

        async fn read(&mut self) -> Result<(f32, f32)> {
            Ok((self.0, self.1))
        }
    }

    // A policy over the pair, commanding a fan duty rather than an on/off.
    struct FanPolicy;

    impl Policy for FanPolicy {
        type Reading = (f32, f32);
        type Command = u8;

        fn evaluate(&mut self, &(temperature, humidity): &(f32, f32)) -> Reaction<u8> {
            let duty = if temperature > 30.0 { 100 } else { 0 };
            Reaction {
                actuator: Some(duty),
                alert: (humidity > 90.0).then_some(Alert::Custom {
                    code: "Muggy",
                    value: humidity,
                }),
            }
        }
    }

    #[tokio::test]
    async fn a_policy_of_your_own_reads_a_pair_and_commands_a_duty() {
        let (mut gateway, link) = connected_pair("farm/#").await;
        let duties = Arc::new(Mutex::new(Vec::new()));
        let mut node = Node::with_policy(
            Profile::irrigation_node(),
            FanPolicy,
            Climate(34.0, 95.0),
            Recording::<u8>(duties.clone()),
            link,
            CborCodec,
        );

        let reaction = node.tick().await.expect("tick");
        assert_eq!(reaction.actuator, Some(100));
        assert_eq!(reaction.alert.map(Alert::kind), Some("Muggy"));
        assert_eq!(*duties.lock().expect("duties"), vec![100]);
        let _: &FanPolicy = node.policy();

        let message = gateway.recv().await.expect("recv").expect("a reading");
        assert_eq!(message.topic, "farm/irrigation/soil-moisture");
        let reading: (f32, f32) = CborCodec.decode(&message.payload).expect("decode");
        assert_eq!(reading, (34.0, 95.0));
    }

    #[tokio::test]
    async fn a_node_runs_from_a_spawned_task() {
        // The generic shape a maker's runner takes: any sensor and actuator, spawned on a
        // multi-threaded runtime, which needs the whole tick to be Send.
        fn spawn_node<S, A, T, C>(mut node: Node<S, A, T, C>) -> tokio::task::JoinHandle<()>
        where
            S: Sensor<Reading = f32> + Send + 'static,
            A: Actuator<Command = bool> + Send + 'static,
            T: Transport + Send + 'static,
            C: Codec<f32> + Send + 'static,
        {
            tokio::spawn(async move {
                for _ in 0..2 {
                    node.tick().await.expect("a tick");
                }
            })
        }

        let (mut gateway, link) = connected_pair("cold-chain/#").await;
        let node = Node::new(
            Profile::vaccine_fridge_monitor(),
            ScriptedSensor::new(vec![9.0, 4.0]),
            Recording(Arc::new(Mutex::new(Vec::new()))),
            link,
            CborCodec,
        );
        spawn_node(node).await.expect("the task finishes");
        let first = gateway.recv().await.expect("recv").expect("a reading");
        let reading: f32 = CborCodec.decode(&first.payload).expect("decode");
        assert_eq!(reading, 9.0);
    }

    #[test]
    fn schedule_follows_state_of_charge() {
        // The schedule reads only the profile, so the components can be placeholders.
        let node = Node::monitor(Profile::vaccine_fridge_monitor(), (), (), ());
        assert_eq!(node.schedule(0.9, false).0, PowerMode::Active);
        assert_eq!(node.schedule(0.1, false).0, PowerMode::Critical);
        // A charging panel eases off by one mode.
        assert_eq!(node.schedule(0.1, true).0, PowerMode::Saver);
        assert_eq!(node.schedule(0.9, false).1, Duration::from_secs(60));
    }
}

File: bindings/dotnet/src/Pamoja.Profile/Node.cs
using System.Globalization;
using System.Text;

using Pamoja.Core;
using Pamoja.Power;

namespace Pamoja.Profile;

/// <summary>
/// What decides each reading: a profile's built-in <see cref="Controller"/>, or a policy of
/// the program's own.
/// </summary>
/// <remarks>
/// A policy of the program's own raises its own conditions as an <see cref="Alert"/> of
/// kind <see cref="AlertKind.Custom"/> with a code and a value.
/// </remarks>
public interface IPolicy
{
    /// <summary>Decides what one reading calls for.</summary>
    /// <param name="reading">The reading, in the unit the profile reads.</param>
    /// <returns>The output setting, if any, and the alert the reading raised.</returns>
    Reaction Evaluate(float reading);
}

/// <summary>
/// What resolves a profile's control kind to the code that decides it: a built-in kind to
/// its <see cref="Controller"/>, and a kind the library never shipped to the policy the
/// factory registered under its name builds.
/// </summary>
/// <remarks>One program then runs any manifest its registry covers.</remarks>
public sealed class PolicyRegistry
{
    private readonly SortedDictionary<string, Func<IReadOnlyDictionary<string, object>, IPolicy>> _factories =
        new(StringComparer.Ordinal);

    /// <summary>Gets the custom kinds registered, in name order.</summary>
    public IReadOnlyList<string> Kinds => [.. _factories.Keys];

    /// <summary>Registers the factory for a custom kind, replacing one of the same name.</summary>
    /// <param name="kind">The kind as a manifest names it, such as <c>frost_guard</c>.</param>
    /// <param name="factory">
    /// Builds the policy from the parameters beside the kind, each a <see cref="double"/>,
    /// a <see cref="bool"/>, or a <see cref="string"/>.
    /// </param>
    /// <returns>The registry, for chaining.</returns>
    public PolicyRegistry Register(string kind, Func<IReadOnlyDictionary<string, object>, IPolicy> factory)
    {
        ArgumentNullException.ThrowIfNull(kind);
        ArgumentNullException.ThrowIfNull(factory);
        _factories[kind] = factory;
        return this;
    }

    /// <summary>Resolves a profile's control kind to the policy that decides it.</summary>
    /// <param name="profile">The profile whose kind is resolved.</param>
    /// <returns>
    /// A fresh policy; a built-in kind's <see cref="Controller"/> is the caller's to dispose.
    /// </returns>
    /// <exception cref="PamojaException">
    /// The kind is custom and no factory is registered for it, naming the kind and the one
    /// it was probably meant to be.
    /// </exception>
    public IPolicy Resolve(Profile profile)
    {
        ArgumentNullException.ThrowIfNull(profile);
        ControlPolicy control = profile.Control;
        if (control.Kind != ControlKind.Custom)
        {
            return profile.Controller();
        }

        string kind = control.CustomKind ?? string.Empty;
        return _factories.TryGetValue(kind, out var factory)
            ? factory(control.Params ?? new Dictionary<string, object>())
            : throw new PamojaException(Nearest.Unresolved(kind, Kinds));
    }
}

/// <summary>One reading a node took, and what its policy decided about it.</summary>
/// <param name="Reading">The reading, in the unit the profile reads.</param>
/// <param name="Reaction">The output setting and the alert the policy decided on.</param>
public readonly record struct Tick(float Reading, Reaction Reaction);

/// <summary>A profile assembled around the parts that make it run.</summary>
/// <remarks>
/// Each <see cref="TickAsync"/> reads, lets the profile's policy decide, switches the
/// output when the policy calls for it, and publishes the reading to the profile's topic.
/// <see cref="RunAsync"/> repeats that at the cadence the profile's power schedule sets for
/// the battery's charge. A node owns the controller it built for the profile, and
/// disposing it releases that.
/// </remarks>
public sealed class Node : IDisposable
{
    private readonly Func<ValueTask<float>> _read;
    private readonly ILink _link;
    private readonly Func<bool, ValueTask>? _drive;
    private readonly IPolicy _policy;
    private readonly bool _ownsPolicy;
    private readonly Func<float, byte[]> _encode;
    private readonly PowerPlan _plan;

    /// <summary>Assembles a node.</summary>
    /// <param name="profile">The profile the node runs.</param>
    /// <param name="read">Takes one reading, in the unit the profile reads.</param>
    /// <param name="link">The link each reading is published over, connected before the node runs.</param>
    /// <param name="drive">Switches the output a profile drives; required for a setpoint profile.</param>
    /// <param name="policy">
    /// What decides each reading: the profile's own controller unless given, a policy of
    /// the program's own, or what <see cref="PolicyRegistry.Resolve"/> returned.
    /// </param>
    /// <param name="encode">Writes a reading as the payload published; the number as JSON text unless given.</param>
    /// <exception cref="PamojaException">
    /// The profile drives an output and no <paramref name="drive"/> is given, or its
    /// control kind is custom and no policy decides it.
    /// </exception>
    public Node(
        Profile profile,
        Func<ValueTask<float>> read,
        ILink link,
        Func<bool, ValueTask>? drive = null,
        IPolicy? policy = null,
        Func<float, byte[]>? encode = null)
    {
        ArgumentNullException.ThrowIfNull(profile);
        ArgumentNullException.ThrowIfNull(read);
        ArgumentNullException.ThrowIfNull(link);
        if (profile.Control.Kind == ControlKind.Setpoint && drive is null)
        {
            throw new PamojaException($"the profile `{profile.Name}` switches an output, so the node needs `drive`");
        }

        Profile = profile;
        _read = read;
        _link = link;
        _drive = drive;
        _ownsPolicy = policy is null;
        _policy = policy ?? profile.Controller();
        _encode = encode ?? (reading => Encoding.UTF8.GetBytes(reading.ToString("R", CultureInfo.InvariantCulture)));
        _plan = profile.PowerPlan;
    }

    /// <summary>Gets the profile the node runs.</summary>
    public Profile Profile { get; }

    /// <summary>Gets the power mode the last <see cref="Schedule"/> chose, or <c>null</c> before the first.</summary>
    public PowerMode? PowerMode { get; private set; }

    /// <summary>Runs one read, decide, act, and publish cycle.</summary>
    /// <returns>The reading and what the policy decided about it.</returns>
    public async Task<Tick> TickAsync()
    {
        float reading = await _read().ConfigureAwait(false);
        Reaction reaction = _policy.Evaluate(reading);
        if (reaction.Actuator is bool on && _drive is not null)
        {
            await _drive(on).ConfigureAwait(false);
        }

        await _link.SendAsync(Profile.Topic, _encode(reading)).ConfigureAwait(false);
        return new Tick(reading, reaction);
    }

    /// <summary>
    /// Says what power mode the battery's charge puts the node in and how long to wait
    /// before the next tick.
    /// </summary>
    /// <remarks>
    /// The node remembers the mode it chose, so a charge hovering at a threshold keeps the
    /// slower cadence until it clears the schedule's hysteresis.
    /// </remarks>
    /// <param name="charge">The state of charge, from 0 to 1.</param>
    /// <param name="charging">Whether the panel is delivering charge.</param>
    /// <returns>The mode, and the wait.</returns>
    public (PowerMode Mode, TimeSpan Wait) Schedule(float charge, bool charging = false)
    {
        PowerMode mode = PowerMode is { } current
            ? _plan.NextModeWhileCharging(current, charge, charging)
            : _plan.ModeWhileCharging(charge, charging);
        PowerMode = mode;
        return (mode, TimeSpan.FromMicroseconds(_plan.IntervalForUs(mode)));
    }

    /// <summary>Ticks, then waits the interval the battery's charge calls for, until cancelled.</summary>
    /// <param name="battery">
    /// Reads the battery before each wait, as a charge from 0 to 1 and whether the panel is
    /// charging; a node without one samples at the active cadence.
    /// </param>
    /// <param name="onTick">Hears each tick, such as to log it or to act on an alert.</param>
    /// <param name="onError">
    /// Hears a tick that failed. The loop carries on at the same cadence; without one, the
    /// failure ends the loop and is thrown.
    /// </param>
    /// <param name="ticks">How many ticks to run before returning; until cancelled unless given.</param>
    /// <param name="wait">Waits between ticks; <see cref="Task.Delay(TimeSpan, CancellationToken)"/> unless given, which a test replaces to run at once.</param>
    /// <param name="cancellationToken">Stops the loop between ticks.</param>
    /// <returns>A task that completes once cancelled or the ticks run out.</returns>
    public async Task RunAsync(
        Func<ValueTask<(float Charge, bool Charging)>>? battery = null,
        Func<Tick, ValueTask>? onTick = null,
        Func<Exception, ValueTask>? onError = null,
        int? ticks = null,
        Func<TimeSpan, CancellationToken, Task>? wait = null,
        CancellationToken cancellationToken = default)
    {
        wait ??= Task.Delay;
        for (int count = 0; ticks is null || count < ticks; count++)
        {
            if (cancellationToken.IsCancellationRequested)
            {
                return;
            }

            try
            {
                Tick tick = await TickAsync().ConfigureAwait(false);
                if (onTick is not null)
                {
                    await onTick(tick).ConfigureAwait(false);
                }
            }
            catch (Exception error) when (onError is not null)
            {
                await onError(error).ConfigureAwait(false);
            }

            (float charge, bool charging) = battery is null ? (1f, false) : await battery().ConfigureAwait(false);
            (_, TimeSpan interval) = Schedule(charge, charging);
            if (ticks is null || count + 1 < ticks)
            {
                try
                {
                    await wait(interval, cancellationToken).ConfigureAwait(false);
                }
                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
                {
                    return;
                }
            }
        }
    }

    /// <inheritdoc/>
    public void Dispose()
    {
        if (_ownsPolicy && _policy is IDisposable owned)
        {
            owned.Dispose();
        }
    }
}
